src/toffy/qc_comp.py

```diff
@@ -545,3 +543,233 @@ def format_img_data(img_data):
+        ).reshape(len(channel_tmas), x_size, y_size)
+
+        # Average the rank for each channel.
+        avg_ranked_tma: np.ndarray = ranked_channels.mean(axis=0)
+
+        cmt_data[qc_col] = avg_ranked_tma
+
+    return cmt_data
+
+
+def batch_effect_qc_metrics(
+    cohort_data_dir: Union[str, pathlib.Path],
+    qc_cohort_metrics_dir: Union[str, pathlib.Path],
+    tissues: List[str],
+) -> None:
+    """
+    Computes QC metrics for a specified set of tissues and saves the tissue specific QC files
+    in the `qc_cohort_metrics_dir`. Calculates the following metrics for the specified tissues,
+    and the metrics for the invidual FOVs within that cohort:
+    * Non-zero mean intensity
+    * Total intensity
+    * 99.9% intensity value
+
+    Args:
+        cohort_data_dir (Union[str, pathlib.Path]): The directory which contains the FOVs for a cohort of interest.
+        qc_cohort_metrics_dir (Union[str,pathlib.Path]): The directory where the cohort metrics will be saved to.
+        tissues (List[str]): A list of tissues to find QC metrics for.
+
+    Raises:
+        ValueError: Errors if `tissues` is either None, or a list of size 0.
+    """
+    if tissues is None or len(tissues) < 1:
+        raise ValueError("The tissues must be specified")
+
+    # Input validation: cohort_data_dir, qc_cohort_metrics_dir
+    io_utils.validate_paths([cohort_data_dir, qc_cohort_metrics_dir])
+
+    samples = io_utils.list_folders(dir_name=cohort_data_dir, substrs=tissues)
+
+    tissue_to_sample_mapping: Dict[str, List[str]] = {}
+
+    for sample in samples:
+        for tissue in tissues:
+            if tissue in sample:
+                if tissue in tissue_to_sample_mapping.keys():
+                    tissue_to_sample_mapping[tissue].append(sample)
+                else:
+                    tissue_to_sample_mapping[tissue] = [sample]
+
+    # Use a set of the samples to avoid duplicate QC metric calculations
+    sample_set = set(list(itertools.chain.from_iterable(tissue_to_sample_mapping.values())))
+
+    # Compute the QC metrics for all unique samples that match with the user's tissue input.
+    for sample in ns.natsorted(sample_set):
+        compute_qc_metrics(
+            extracted_imgs_path=cohort_data_dir, fov_name=sample, save_csv=qc_cohort_metrics_dir
+        )
+
+    # Combined metrics per Tissue
+    for tissue, samples in tissue_to_sample_mapping.items():
+        for ms in settings.QC_SUFFIXES:
+            metric_files: List[str] = io_utils.list_files(
+                qc_cohort_metrics_dir, substrs=[f"{sample}_{ms}.csv" for sample in samples]
+            )
+
+            metric_files = list(filter(lambda mf: "combined" not in mf, metric_files))
+
+            # Define an aggregated metric DataFrame
+            combined_metric_tissue_df: pd.DataFrame = pd.concat(
+                (pd.read_csv(os.path.join(qc_cohort_metrics_dir, mf)) for mf in metric_files)
+            )
+
+            combined_metric_tissue_df.to_csv(
```

Comment:
assuming these files will get saved where the mph and pulse heights files currently are?

Reply:
I was thinking a different directory, but let me know if it makes more sense for it to be in the same one.

Reply:
I guess they can probably go wherever, since this isn't being implemented in watcher!